symfit/core/fit.py

```diff
@@ -1057,30 +1057,30 @@ def covariance_matrix(self, best_fit_params):
         """
         cov_matrix = self._covariance_matrix(best_fit_params,
                                              objective=self.objective)
-        if cov_matrix is None:
-            # If the covariance matrix could not be computed, we try again by
-            # approximating the hessian with the jacobian.
-            if not isinstance(self.objective, HessianObjective) \
-                    or not isinstance(self.model, HessianModel):
-                # VectorLeastSquares should be turned into a LeastSquares for
-                # cov matrix calculation
-                if self.objective.__class__ is VectorLeastSquares:
-                    base = LeastSquares
-                else:
-                    base = self.objective.__class__
-
-                class HessApproximation(base, HessianObjectiveJacApprox):
-                    """
-                    Class which impersonates ``base``, but which returns zeros
-                    for the models Hessian. This will effectively result in the
-                    calculation of the approximate Hessian by calculating
-                    outer(J.T, J) when calling ``base.eval_hessian``.
-                    """
-
-                objective = HessApproximation(self.objective.model,
-                                              self.objective.data)
-                cov_matrix = self._covariance_matrix(best_fit_params,
-                                                     objective=objective)
+        if cov_matrix is None or any(np.diag(cov_matrix) < 0):
+            # If the covariance matrix could not be computed or contains illegal
+            # values, we try again by approximating the hessian with the
+            # jacobian.
```

Comment:
More importantly, how/why/when does this happen? And how negative are those values?

Reply:
I agree that it is weird, we are not talking about small value here. It happens for `TestFitResults.test_fitting_2`, and the variances can be as big as `-1150`. Since this means an imaginary standard deviation, the quickest fix is to do the Jacobian approximation, which is guaranteed to be positive-definite.

I'm not sure why it happens, maybe it's something like this: https://stats.stackexchange.com/questions/30465/what-does-a-non-positive-definite-covariance-matrix-tell-me-about-my-data.

Maybe this test is poorly posed to begin with, though it is not an uncommon situation.

Reply:
p.s. I've now made all the values in this test less ideal so the components become independent, but the problem persists.

Reply:
Hmmnn.. To summarize the SE post: there are colinear Parameters? This should at the very least raise a *loud* warning to the user that something is fishy. We could even tell them which parameters are colinear; either automagically, or expose a function/method that tells them.

Reply:
Yeah, maybe it is actually better to change our tests to accept NaN as the proper answer instead of trying to give the old result which is due to a (standard) approximation. The model used in the mentioned test does fail to include the covariance, so perhaps this is indeed due to a poor model. And [here ](https://www.researchgate.net/post/What_is_the_explanation_of_negative_variance_in_the_statistical_models)it is also mentioned that software should be honest about negative covariances, with which I completely agree. Let's just confront users with it so they can conclude for themselves what it means.

Reply:
A better guess seems to have solved the problem, but I also stand by my previous comment.